src/odemis/util/comp.py

```diff
@@ -113,64 +115,67 @@ def get_fov_rect(comp, fov):
+    :param zstep: distance between two successive zlevels. If negative,
+    the order of the returned list of zlevels is reversed. The actual zstep in the
+    returned list is adjusted to the closest value of the given zstep that divides
+    the zrange. If zstep is too small, this will lead to too large number of zlevels.
+    In that case IndexError is raised.
+    :returns: list of zlevels, where each zlevel is absolute position for
+    the focuser.
     """
     if zstep == 0:
-        raise ZeroDivisionError("The step size 'Zstep' can not be zero")
+        raise ZeroDivisionError("The step size 'zstep' can not be zero")
     if zrange[0] > zrange[1]:
-        raise ValueError("The given range is not correct. zmin should be first and zmax second")
+        raise ValueError(f"The given range {zrange} is not correct. The first value should be smaller than the second one.")
     if "z" not in focuser.axes.keys():
-        raise KeyError("The focus actuator %s does not have z axis" %focuser)
+        raise KeyError(f"The focus actuator {focuser} does not have z axis")
 
     focuser_pos = focuser.position.value["z"]
+    zrange_abs = (zrange[0] + focuser_pos, zrange[1] + focuser_pos)
 
     # Get the range from the axis range + extra limit POS_ACTIVE_RANGE
     focuser_rng = focuser.axes["z"].range
     sw_rng = focuser.getMetadata().get(model.MD_POS_ACTIVE_RANGE)
-    if sw_rng:
-        focuser_rng = (max(focuser_rng[0], sw_rng[0]),
-                       min(focuser_rng[1], sw_rng[1]))
-
-    # clip the zMax and zMin to the actuator limits if necessary 
-    if zrange[1] + focuser_pos > focuser_rng[1]:
-        zrange[1] = focuser_rng[1] - focuser_pos
-    if focuser_pos + zrange[0] < focuser_rng[0]:
-        zrange[0] = focuser_rng[0] - focuser_pos
-    if zrange[0] == zrange[1]:
+    if sw_rng and "z" in sw_rng:
+        focuser_rng = (max(focuser_rng[0], sw_rng["z"][0]),
+                       min(focuser_rng[1], sw_rng["z"][1]))
+
+    # clip the zMax and zMin to the actuator limits if necessary
+    zrange_abs = (min(max(focuser_rng[0], zrange_abs[0]), focuser_rng[1]),
+                  min(max(focuser_rng[0], zrange_abs[1]), focuser_rng[1]))
+
+    if zrange_abs[0] == zrange_abs[1]:
         return focuser_pos + zrange[0]
 
     # find number of samples 
-    n = (zrange[1] - zrange[0]) / abs(zstep) + 1
+    n = (zrange_abs[1] - zrange_abs[0]) / abs(zstep) + 1
     if n > MAX_ZLEVELS:
-        raise IndexError("The number of zlevels, %s, is too large. The max allowable number of zlevels is %s. Reduce the zstep value." %(n, MAX_ZLEVELS))
+        raise IndexError(f"The number of zlevels, {n}, is too large. Reduce the zstep value to < {MAX_ZLEVELS}.")
+
     # try the floor and ceil values for the number of samples, and 
     # take the one that give smaller error.
     # note: if n is 1 -> division by zero and step_c will always be picked,
     # so just choose any large value for step_f (e.g infinite)
```

Comment:
this comment is incorrect, if n is exactly 1 there will also be a division by zero error for the step_c, which is then not handled at all.

Reply:
you could also write it like this:
```python
    n_steps_f = math.floor(n) - 1
    n_steps_c = math.ceil(n) - 1

    if n_steps_f == 0:
        n = math.ceil(n)
    else:
        # errors
        ef = abs((zrange_abs[1] - zrange_abs[0]) / n_steps_f - abs(zstep))
        ec = abs((zrange_abs[1] - zrange_abs[0]) / n_steps_c - abs(zstep))
        n = math.floor(n) if ef < ec else math.ceil(n)
```

Reply:
I think that n can never be exactly 1 because it's defined as `(zrange_abs[1] - zrange_abs[0]) / abs(zstep) + 1` and we already handle the case of zrange_abs being null. 
Anyway, I think your version is clearer. I've adjusted it a tiny bit more.... and clarified the comments.